### DISCUSSION TITLE
Fill, stroke fail in -draw command

### DISCUSSION THREAD

**Discussion by SkepticaLee** [ImageMagick/ImageMagick#4205] [Help]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched [open](https://github.com/ImageMagick/ImageMagick/issues) and [closed](https://github.com/ImageMagick/ImageMagick/issues?q=is%3Aissue+is%3Aclosed) issues to ensure it has not already been reported
- [X] I have verified that I am using the latest version of ImageMagick

### ImageMagick version

7.0.11-10

### Operating system

Windows

### Operating system, version and so on

Windows 10, from the console

### Description

When drawing bezier curves with other objects in one -draw command, only the first set of "fill", "fill-opacity", "stroke", "stroke-opacity", "stroke-width" assignments are implemented, a further set results in no drawing

### Steps to Reproduce

This just produces four red circles on a green background (script 1, image 1):
magick -size 100x100 xc:limegreen -draw "fill red fill-opacity 1 stroke-width 0 stroke black stroke-opacity 0 circle 20 20 20 22 circle 30 30 30 32 circle 70 50 70 52 circle 80 20 80 22 fill black fill-opacity 0 stroke-width 2 stroke blue stroke-opacity 1 bezier 20 20 -3.5 24.7 25.1 27.1 30 30 bezier 30 30 34.8 32.9 52.4 53.5 70 50 bezier 70 50 87.5 46.5 90.8 26.5 80 20 bezier 80 20 69.2 13.5 43.5 15.3 20 20" test.png

However this produces the line with the same colors as for the circles meaning that the bezier curves are partially filled (script 2, image 2), which is not the desired effect (image 3):
magick -size 100x100 xc:limegreen -draw "fill red fill-opacity 1 stroke-width 1 stroke black stroke-opacity 1 circle 20 20 20 22 circle 30 30 30 32 circle 70 50 70 52 circle 80 20 80 22 bezier 20 20 -3.5 24.7 25.1 27.1 30 30 bezier 30 30 34.8 32.9 52.4 53.5 70 50 bezier 70 50 87.5 46.5 90.8 26.5 80 20 bezier 80 20 69.2 13.5 43.5 15.3 20 20" test.png

### Images

First script:
![test](https://user-images.githubusercontent.com/24591664/118869808-248ad800-b8e6-11eb-9ed1-85d44b48f773.png)

Second script:
![test](https://user-images.githubusercontent.com/24591664/118869876-3a000200-b8e6-11eb-9457-233c47cb83f1.png)

Desired effect:
![test](https://user-images.githubusercontent.com/24591664/118870065-72074500-b8e6-11eb-9919-9f034d6f948a.png)

**Comment by fmw42** [comment]
You need fill red only for the circles. So set fill none just before the bezier curves are drawn. This is not a bug, but the way bezier curves are drawn.

`magick -size 100x100 xc:limegreen -draw "fill red fill-opacity 1 stroke-width 1 stroke black stroke-opacity 1 circle 20 20 20 22 circle 30 30 30 32 circle 70 50 70 52 circle 80 20 80 22 fill none bezier 20 20 -3.5 24.7 25.1 27.1 30 30 bezier 30 30 34.8 32.9 52.4 53.5 70 50 bezier 70 50 87.5 46.5 90.8 26.5 80 20 bezier 80 20 69.2 13.5 43.5 15.3 20 20" test.png`

![test](https://user-images.githubusercontent.com/13458218/118871059-0ef0c080-b89c-11eb-9f64-d0f914e88748.png)

**Comment by SkepticaLee** [comment]
Sorry, Fred, but the bug is why I can't set the colors using full descriptions (color, width, opacity)? Surely that ought to work?

**Comment by fmw42** [comment]
Sorry, I do not understand. I do not see that as a bug. That is how bezier curves work.  Perhaps I misunderstand. So please elaborate. Why can you not do what I did by resetting the fill color?

**Comment by snibgo** [comment]
@SkepticaLee: I suggest you split your "-draw" operation into parts, so you have one "-draw" that sets colours and draws circles, then another "-draw" for setting colours and drawing the beziers.

**Comment by SkepticaLee** [comment]
Fred, we'll leave aside the beziers for a bit.  The issue is that setting stroke opacity to 0 once makes setting the color, opacity or width again moot.

```
magick -size 100x100 xc:#BF8\
-draw "fill orange fill-opacity 1 stroke blue stroke-width 2 stroke-opacity 0.5 circle 25 25 25 50\
fill white fill-opacity 0 stroke red stroke-width 3 stroke-opacity 1 circle 70 70 70 95" test_0.png
```
 
![test_0](https://user-images.githubusercontent.com/24591664/118965530-1a5ded80-b969-11eb-96eb-036db96791eb.png)

Works as expected. But

```
magick -size 100x100 xc:#BF8\
-draw "fill orange fill-opacity 0.5 stroke blue stroke-width 2 stroke-opacity 0 circle 25 25 25 50\
fill white fill-opacity 0.5 stroke red stroke-width 3 stroke-opacity 1 circle 70 70 70 95" test_1.png
```

![test_1](https://user-images.githubusercontent.com/24591664/118965640-2ea1ea80-b969-11eb-9baf-5c84fa4a5415.png)

Produces two circles without any stroke. (By the way circles should need just three parameters, not four.)

If you are doing this only once in a lifetime, then you could manually change the first stroke setting to "none" and the second one works. But what if I were producing hundreds of images for animations, and couldn't predict where "stroke-opacity 0" would turn up? This caveat also applies to snibgo's suggestion of splitting the -draw commands. Again, if you're only doing this once, then you can inspect the command and try splitting it up, but that's not practical for an automated process. If you had to create a new -draw command every time something happened to stroke-opacity, that would involve determining the values for stroke (color) and stroke-width and carrying them forward every time. Not very user friendly.

All I want is that a call to stroke-opacity 0 to be undone by a subsequent call to stroke-opacity > 0 in the same -draw command. Seems to work for fill-opacity.

**Comment by snibgo** [comment]
See my comment above. For example (bash script):
```
magick.exe -size 100x100 xc:#BF8 \
-draw "fill orange fill-opacity 0.5 stroke blue stroke-width 2 stroke-opacity 0 circle 25 25 25 50" \
-draw "fill white fill-opacity 0.5 stroke red stroke-width 3 stroke-opacity 1 circle 70 70 70 95" \
test_2.png
```

**Comment by SkepticaLee** [comment]
Snibgo, I did comment on your workaround. But it is just that. It is not practicable in an automated setting.

The bug is that fill/stroke-opacity 0 causes all further calls to fill/stroke-anything to be fruitless.

**Comment by fmw42** [comment]
Sounds like a bug to me. I will leave it to the IM developers to comment further.

**Comment by SkepticaLee** [comment]
Another, perhaps related curiosity is when setting both the fill color to none, and the fill-opacity to 0, the fill color is actually black.

As far as I can see this change in behavior took place somewhere between versions 7.0.7 and the current one.

```
magick -size 100x100 xc:#FBB -draw "fill none fill-opacity 0 circle 30 30 60 30" test_0.png
```
![test_0](https://user-images.githubusercontent.com/24591664/119020821-fe753e80-b99e-11eb-9a9c-57c0849a62c1.png)

**Comment by fmw42** [comment]
When you set fill-opacity to 1, you get a fully transparent circle. That is it is invisible and all you see is pink.

At some point in IM 7, opacity and transparency got reversed. Some things may not have been properly reconfigured. Perhaps this is one.

Also I am puzzled by your using a transparent fill color and having opacity control. Seems like they are competing processes.

`magick -size 100x100 xc:"#FBB" -draw "fill none fill-opacity 1 circle 30 30 60 30" test_0.png`

![test_0](https://user-images.githubusercontent.com/13458218/119022142-1fb13d00-b955-11eb-8676-f67ca8414ad1.png)

You cannot draw a "hole" with MSVG. You can flood fill a "hole". But I do not know how you can directly draw a hole.

**Comment by SkepticaLee** [comment]
Right on, Fred. If you use any value for fill-opacity, you get a pink square, except for 0. Which should also give you a pink square.

**Comment by fmw42** [comment]
I tried 0.5 for the fill-opacity and did not get a fully pink square. There was a grayish circle. So the black was half blended with pink.

**Comment by SkepticaLee** [comment]
If the fill is black and opacity 0, then no circle shows. I expected the same behavior for "none", but that is not the case. In fact, it shouldn't matter what the opactiy value is if the color is "none" (the behavior of v 7.0.7).

```
magick -size 100x100 xc:#FBB -draw "fill black fill-opacity 0 circle 30 30 60 30" test_0.png
```
![test_0](https://user-images.githubusercontent.com/24591664/119029686-3bdec980-b9a9-11eb-9ac0-a90e7bafb068.png)

**Comment by fmw42** [comment]
As I said, there seems to be competing processed: fill color of none and opacity. Perhaps the opacity takes precedence, since you cannot draw (fill none) a "hole". So effectively the fill none become fill black.
